### PR TITLE
Allow `impl Trait` inside where clauses

### LINKED ISSUE #104526: Confusing `impl Trait` limitation
For some reason, `impl Trait` is allowed in a type's first mention but not in a `where` clause.

```rust
use core::ops::Add;

// Allowed
fn foo<T: Add<Output = impl Default>>(t: T) {}
fn bar<T: AsRef<impl Default>>(t: T) {}

// Not allowed
fn another_foo<T>(t: T)
where
    T: Add<Output = impl Default>,
{
}
fn another_bar<T>(t: T)
where
    T: AsRef<impl Default>,
{
}
```

Besides, the error message isn't very helpful.

```
`impl Trait` only allowed in function and inherent method return types, not in *****bound***** (which bound?)
```

### PR DESCRIPTION
**Pull request by bombless**
Fix https://github.com/rust-lang/rust/issues/104526

### CODE REVIEW

File: compiler/rustc_error_messages/locales/en-US/ast_lowering.ftl [hunk 1]
```diff
@@ -17,7 +17,7 @@ ast_lowering_assoc_ty_parentheses =
 ast_lowering_remove_parentheses = remove these parentheses
 
 ast_lowering_misplaced_impl_trait =
-    `impl Trait` only allowed in function and inherent method return types, not in {$position}
+    `impl Trait` not allowed within {$position}
```
Comment (estebank) [inline compiler/rustc_error_messages/locales/en-US/ast_lowering.ftl:20]:
```suggestion
    `impl Trait` isn't allowed within {$position}
```

File: src/test/ui/impl-trait/where-allowed.rs [hunk 2]
```diff
@@ -154,97 +155,111 @@ type InTypeAlias<R> = impl Debug;
...
 struct InInherentImplAdt<T> { t: T }
 impl InInherentImplAdt<impl Debug> {
-    //~^ ERROR `impl Trait` only allowed in function and inherent method return types
+    //~^ ERROR `impl Trait` not allowed within type [E0562]
 }
 
-// Disallowed in where clauses
+// Allowed in where clauses
 fn in_fn_where_clause()
     where impl Debug: Debug
```
Comment (estebank) [inline src/test/ui/impl-trait/where-allowed.rs:184]: Wait, I'm not sure this kind of bounds being allowed make much sense. What is the usecase for this?

Comment (bombless) [reply src/test/ui/impl-trait/where-allowed.rs:184]: It's just like `fn f<T: Debug>() where T: Debug {}`.
I guess it's best to keep it this way to make it consistent.

File: compiler/rustc_error_messages/locales/en-US/ast_lowering.ftl (hunk 1 above)
Comment (bombless) [reply compiler/rustc_error_messages/locales/en-US/ast_lowering.ftl]: done.

File: src/test/ui/impl-trait/where-allowed.rs (hunk 2 above)
Comment (compiler-errors) [reply src/test/ui/impl-trait/where-allowed.rs:184]: but unlike `fn f<T: Debug>`, given that this is a universal/"argument-position" `impl Trait`, neither the caller nor the callee can *use* this `impl Trait` (caller can't provide it, callee can't reference), so it's particularly limited.

Comment (bombless) [reply src/test/ui/impl-trait/where-allowed.rs:184]: Yes that's a problem though.
I'll add a lint for it.
Also, what if we need a type somewhere that related to a trait, but we don't care what is the exact type, e.g. `fn f<T>(v: T) where Result<impl Default, T>: Default {}`. I think it's also a valid use-case, even though it's a by-product of impl-trait.

**Comment by nagisa** [comment]
r? rust-lang/compiler

**Comment by compiler-errors** [comment]
r? @compiler-errors 
@rustbot author

@bombless what's the status on this? Can you rebase?

**Comment by bombless** [comment]
I don't have time to finish the patch currently, so sorry I'm closing this PR

**Pull request closed without merging**